### PR TITLE
Add dremio authorization by calling apiv2/login

### PR DESCRIPTION
**Pull request by Sevenannn**
Call dremio http://{hostname}/apiv2/login endpoint to verify username and password when user call `spice login dremio`
The hostname is currently hard-coded to be the same as cli/cmd/dataset.go dremio endpoint.

Something we can improve in the following PRs:
* Allow user to specify dremio host in `spice login dremio`.
* Use dremio authentication token (issued at dremio authentication, expire in 30 hours) for further interactions with dremio (e.g. get data).

### CODE REVIEW

File: bin/spice/pkg/cli/cmd/login.go [hunk 1]
```diff
@@ -150,6 +157,51 @@ spice login dremio --username <username> --password <password>
...
+func dremioAuth(username string, password string) bool {
+
+	type userCredentials struct {
+		UserName string `json:"userName"`
+		Password string `json:"password"`
+	}
+
+	// TODO: Allow user to specify their own dremio host. Needs UX design for how the command should handle.
+	// TODO: Do we ultimately want to use dremio token instead of username/password in the following dremio requests?
+	const dremioHost = "dremio-4mimamg7rdeve.eastus.cloudapp.azure.com:9047"
```
Comment (phillipleblanc) [inline bin/spice/pkg/cli/cmd/login.go:169]: Let's remove this hard-coding, ref: https://github.com/spicehq/spiceai/issues/100

**Review by phillipleblanc — CHANGES_REQUESTED** [review]
We decided not to put this into the current release

File: bin/spice/pkg/cli/cmd/login.go [hunk 2]
```diff
@@ -150,6 +157,51 @@ spice login dremio --username <username> --password <password>
 	},
 }
 
+func dremioAuth(username string, password string) bool {
+
+	type userCredentials struct {
+		UserName string `json:"userName"`
```
Comment (lukekim) [inline bin/spice/pkg/cli/cmd/login.go:163]: Weird casing. Don't capitalize N.

File: bin/spice/pkg/cli/cmd/login.go [hunk 3]
```diff
@@ -138,6 +140,11 @@ spice login dremio --username <username> --password <password>
 			os.Exit(1)
 		}
 
+		if !dremioAuth(username, password) {
+			cmd.Println("Invalid dremio username or password")
```
Comment (lukekim) [inline bin/spice/pkg/cli/cmd/login.go:144]: Capital D for Dremio.

Comment (lukekim) [reply bin/spice/pkg/cli/cmd/login.go:144]: Actually, just remove the word Dremio altogether.

File: bin/spice/pkg/cli/cmd/login.go [hunk 4]
```diff
@@ -150,6 +157,51 @@ spice login dremio --username <username> --password <password>
...
+
+	type userCredentials struct {
+		UserName string `json:"userName"`
+		Password string `json:"password"`
+	}
+
+	// TODO: Allow user to specify their own dremio host. Needs UX design for how the command should handle.
+	// TODO: Do we ultimately want to use dremio token instead of username/password in the following dremio requests?
+	const dremioHost = "dremio-4mimamg7rdeve.eastus.cloudapp.azure.com:9047"
+	dremioLoginAddr := fmt.Sprintf("http://%s/apiv2/login", dremioHost)
```
Comment (lukekim) [inline bin/spice/pkg/cli/cmd/login.go:170]: This should use https if supported.

File: bin/spice/pkg/cli/cmd/login.go [hunk 5]
```diff
@@ -150,6 +157,51 @@ spice login dremio --username <username> --password <password>
...
+
+	creds := userCredentials{
+		UserName: username,
+		Password: password,
+	}
+
+	jsonData, err := json.Marshal(creds)
```
Comment (lukekim) [inline bin/spice/pkg/cli/cmd/login.go:177]: This won't work without attributes on the struct.

File: bin/spice/pkg/cli/cmd/login.go [hunk 6]
```diff
@@ -150,6 +157,51 @@ spice login dremio --username <username> --password <password>
...
+	if err != nil {
+		fmt.Println("Error marshaling JSON:", err)
+	}
+
+	req := fasthttp.AcquireRequest()
+	defer fasthttp.ReleaseRequest(req)
+
+	req.SetRequestURI(dremioLoginAddr)
+	req.Header.SetMethod("POST")
+	req.Header.SetContentType("application/json")
```
Comment (lukekim) [inline bin/spice/pkg/cli/cmd/login.go:187]: Do we not have a help for a JSON request?

File: bin/spice/pkg/cli/cmd/login.go [hunk 7]
```diff
@@ -150,6 +157,51 @@ spice login dremio --username <username> --password <password>
...
+	err = fasthttp.Do(req, resp)
+	if err != nil {
+		fmt.Println("Error sending request:", err)
+	}
+
+	if resp.StatusCode() == fasthttp.StatusOK {
+		return true
+	}
+
+	return false
```
Comment (lukekim) [inline bin/spice/pkg/cli/cmd/login.go:202]: Don't return a bool. Align to Go conventions. Return an error if it fails and nil if it succeeds. or both a bool and error.

**Pull request closed without merging**